doom/.doom.d/config.org
#+TITLE: Doom Emacs Config
#+AUTHOR: Alexander Brevig
#+EMAIL: alexanderbrevig@gmail.com
#+LANGUAGE: en
#+PROPERTY: header-args :tangle yes

* Table of Contents :TOC_4_gh:
- [[#basic-doom-config][Basic Doom Config]]
  - [[#lexical-binding][Lexical Binding]]
  - [[#personal-information][Personal Information]]
- [[#ui-settings][UI Settings]]
  - [[#fonts][Fonts]]
  - [[#default-theme][Default Theme]]
  - [[#basic-ui-settings][Basic UI Settings]]
- [[#ux-settings][UX Settings]]
  - [[#basic-ux-settings][Basic UX Settings]]
  - [[#stop-confirm-kill][Stop Confirm Kill]]
  - [[#splits-that-works][Splits That Works]]
    - [[#split-direction][Split Direction]]
    - [[#prompt-buffer-on-splits][Prompt Buffer On Splits]]
  - [[#responsive][Responsive]]
  - [[#projectile][Projectile]]
- [[#lsp][LSP]]
  - [[#common-lisp-slime][Common Lisp slime]]
- [[#org][Org]]
  - [[#directory-config][Directory Config]]
  - [[#org-roam][Org Roam]]
    - [[#v2-ftw][v2 FTW]]
    - [[#capture-templates][Capture Templates]]
      - [[#consider-customizing-cap-templ-for-org-roam][Consider customizing cap-templ for org-roam]]
  - [[#autosync-with-git][Autosync with git]]

* Basic Doom Config
** Lexical Binding
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src
** Personal Information
#+begin_src emacs-lisp
(setq user-full-name "Alexander Brevig"
      user-mail-address "alexanderbrevig@gmail.com")
#+end_src
* UI Settings
** Fonts
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 16)
      doom-big-font (font-spec :family "FiraCode Nerd Font" :size 22)
      doom-variable-pitch-font (font-spec :family "Overpass Nerd Font" :size 14))
#+end_src
** Default Theme
#+begin_src emacs-lisp
(setq doom-theme 'doom-gruvbox)
#+end_src
** Basic UI Settings
#+begin_src emacs-lisp
(setq doom-modeline-mode t)

(setq display-line-numbers-type 'relative)
#+end_src
* UX Settings
** Basic UX Settings
#+begin_src emacs-lisp
(setq scroll-margin 5)
#+end_src
** Stop Confirm Kill
#+begin_src emacs-lisp
(setq confirm-kill-emacs 'nil)
#+end_src
** Splits That Works
*** Split Direction
#+begin_src emacs-lisp
(setq evil-vsplit-window-right t
      evil-split-window-below t)
#+end_src
*** Prompt Buffer On Splits
#+begin_src emacs-lisp
(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))
#+end_src
** Responsive
#+begin_src emacs-lisp
(setq which-key-idle-delay 0)
** Projectile
#+begin_src emacs-lisp
(setq projectile-project-search-path '("~/entur/bb" "~/entur/gh" "~/entur/mat" "~/github.com" "~/gitlab.com"))
#+end_src
* LSP
#+begin_src emacs-lisp
(flycheck-mode 1)
(format-all-mode 1)
#+end_src
** Common Lisp slime
#+begin_src emacs-lisp
(load (expand-file-name "~/.quicklisp/slime-helper.el"))
(setq inferior-lisp-program "sbcl")
#+end_src
* Org
** Directory Config
#+begin_src emacs-lisp
(setq org-directory "~/github.com/org/")
(after! org
  (setq org-agenda-files (list "~/github.com/dotfiles/doom/.doom.d/config.org"
                               "~/github.com/org/")))
(setq org-roam-directory "~/github.com/org-roam/")
#+end_src
** Org Roam
*** v2 FTW
#+begin_src emacs-lisp
(setq org-roam-v2-ack t)

(use-package! org-roam
  :after org
  :config
  (setq org-roam-v2-ack t)
  (setq org-roam-mode-sections
        (list #'org-roam-backlinks-insert-section
              #'org-roam-reflinks-insert-section
              #'org-roam-unlinked-references-insert-section))
  (org-roam-setup))
#+end_src

*** Capture Templates

**** TODO Consider customizing cap-templ for org-roam
** Autosync with git
#+begin_src emacs-lisp
(defun org-git-sync (repo)
  (interactive)
  (if (s-contains? (format "github.com/%s" repo) (buffer-file-name))
      (progn
        (magit-with-toplevel (magit-stage-1 nil (list buffer-file-name)))
        (magit-with-toplevel (magit-run-git
                              "commit" "-m"
                              (format "%s <%s>"
                                      (file-name-nondirectory (buffer-file-name))
                                      (format-time-string "%Y-%m-%dT%T"))))
        (magit-with-toplevel (magit-run-git "push")))))

(defun org-git-sync-org ()
  (interactive)
  (org-git-sync "org"))

(defun org-git-sync-org-roam ()
  (interactive)
  (org-git-sync "org-roam"))

(add-hook 'after-save-hook 'org-git-sync-org)
#+end_src


emacs/.emacs.d/config.org
* Emacs
** Fix UI
#+BEGIN_SRC emacs-lisp
  (when window-system
    (global-hl-line-mode t)
    (global-prettify-symbols-mode t))
  (global-display-line-numbers-mode 1)
  (setq display-line-numbers 'relative)
#+END_SRC
** Fix UX
#+begin_src emacs-lisp
  (setq scroll-conservatively 100)
  (setq inhibit-startup-message t)
  #+end_src
** Fix files
#+begin_src emacs-lisp
  (setq make-backup-file nil)
  (setq auto-save-default nil)
#+end_src
* Packages
** use-package
#+begin_src emacs-lisp
  (require 'package)
  (setq package-enable-at-startup nil)
  (add-to-list 'package-archives
	       '("melpa" . "https://melpa.org/packages/"))

  (package-initialize)

  (unless (package-installed-p 'use-package)
    (package-refresh-contents)
    (package-install 'use-package))
#+end_src
** General.el
#+begin_src emacs-lisp
  (use-package general
    :ensure t)

#+end_src
** EVIL
*** evil
#+begin_src emacs-lisp
  (use-package evil
    :ensure t
    :init (evil-mode))
#+end_src
*** evil-org
#+begin_src emacs-lisp
  (use-package evil-org
    :ensure t
    :after org
    :hook (org-mode . (lambda () evil-org-mode))
    :config
    (require 'evil-org-agenda)
    (evil-org-agenda-set-keys))
#+end_src
** Evil keymaps
*** leader
#+begin_src emacs-lisp
  (general-def :states '(normal motion emacs) "SPC" nil)

  (general-create-definer my-leader
    :prefix "SPC")
  (evil-set-leader nil (kbd "SPC"))
  (evil-set-leader 'normal (kbd "SPC"))
#+end_src
*** mode jumps
#+begin_src emacs-lisp
  (my-leader 'normal 'override
    "o" '(:ignore t :which-key "org")
    "oa" 'org-agenda
    "oc" 'org-capture)
#+end_src
*** escape
#+begin_src emacs-lisp
  (define-key evil-insert-state-map (kbd "C-c C-c") 'evil-normal-state)
  (define-key evil-normal-state-map (kbd "C-c C-c") 'evil-normal-state)
#+end_src
*** lspkeys
#+begin_src emacs-lisp
#+end_src
** Looks
#+begin_src emacs-lisp
  (use-package doom-modeline
    :ensure t
    :init (doom-modeline-mode 1))

  (use-package doom-themes
    :ensure t
    :config
    (setq doom-themes-enable-bold t
	  doom-themes-enable-italics t)
    (load-theme 'doom-gruvbox t)
    (doom-themes-visual-bell-config)
    (doom-themes-org-config))
#+end_src
** Utils
#+begin_src emacs-lisp
  (use-package which-key
    :ensure t
    :init (which-key-mode))
#+end_src
*** Shell
#+begin_src emacs-lisp
  (defalias 'yes-or-no-p 'y-or-n-p)
  (defvar my-term-shell "/usr/bin/zsh")
  (defadvice ansi-term (before force-bash)
    (interactive (list my-term-shell)))
  (ad-activate 'ansi-term)

  (my-leader 'normal 'override
    "t" 'ansi-term)

  (defun set-no-process-query-on-exit ()
    (let ((proc (get-buffer-process (current-buffer))))
      (when (processp proc)
	(set-process-query-on-exit-flag proc nil))))

  (add-hook 'term-exec-hook 'set-no-process-query-on-exit)
#+end_src
*** IDO
#+begin_src emacs-lisp
  (setq ido-enable-flex-matching nil)
  (setq ido-create-new-buffer 'always)
  (setq ido-everywhere t)
  (ido-mode 1)
#+end_src
*** Config management
#+begin_src emacs-lisp
  (defun config-visit ()
    (interactive)
    (find-file "~/.emacs.d/config.org"))
  (global-set-key (kbd "C-c e") 'config-visit)
#+end_src
#+begin_src emacs-lisp
  (defun config-reload ()
    (interactive)
    (org-babel-load-file (expand-file-name "~/.emacs.d/config.org")))
  (global-set-key (kbd "C-c r") 'config-reload)
#+end_src

* LSP
** Flycheck
#+begin_src emacs-lisp
  (use-package flycheck :ensure t)
#+end_src
** Yasnippet
#+begin_src emacs-lisp
  (use-package yasnippet
    :ensure t
    :config (setq yas-verbosity 1 yas-wrap-around-region t)

    (with-eval-after-load 'yasnippet
      (setq yas-snippet-dirs '(yasnippet-snippets-dir)))
    (yas-reload-all)
    (yas-global-mode))
  (use-package yasnippet-snippets
    :ensure t)
#+end_src
** LSP mode
#+begin_src emacs-lisp
  (use-package lsp-mode
    :ensure t
    :init
    (setq lsp-keymap-prefix "<leader>l")
    :config
    (evil-define-key 'normal lsp-mode-map (kbd "<leader>l") lsp-command-map)
    :hook (
	   (go-mode . lsp)
	   ;;
	   (lsp-mode . lsp-enable-which-key-integration))
    :commands lsp)
  (use-package lsp-ui
    :ensure t
    :init (setq lsp-ui-sideline-enable t)
    :commands lsp-ui-sideline-mode)


  ;;(use-package lsp-ivy :commands lsp-ivy-workspace-symbol)
#+end_src
** Company code completion
#+begin_src emacs-lisp
  (use-package company
    :ensure t
    :bind (:map company-active-map
		("C-j" . company-select-next)
		("C-k" . company-select-previous))
    :config
    (setq company-ide-delay 0)
    (setq company-minimum-previs-length 1)
    (setq company-tooltip-align-annotations t))
  (company-mode 1)
#+end_src
** Go mode
#+begin_src emacs-lisp
  (use-package go-mode
    :ensure t
    :hook ((go-mode . lsp-deferred)
      (before-save . lsp-format-buffer)
      (before-save . lsp-organize-imports)))
#+end_src


doom/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(doom-load-envvars-file "~/.doom.d/myenv")

(setq user-full-name "Alexander Brevig"
      user-mail-address "alexanderbrevig@gmail.com")

;; Doom config
(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 16)
      doom-big-font (font-spec :family "FiraCode Nerd Font" :size 22)
      doom-variable-pitch-font (font-spec :family "Overpass Nerd Font" :size 14))

(setq doom-theme 'doom-gruvbox)

(setq doom-modeline-mode t)

(setq display-line-numbers-type 'relative)

(setq org-directory "~/github.com/org")
(setq org-roam-directory "~/github.com/org-roam")

(setq confirm-kill-emacs 'nil
      evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))

(setq scroll-margin 5)

;; LSP and check
(flycheck-mode 1)
(format-all-mode 1)
